### PR TITLE
Rework .NET samples and snippets instructions

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes dotnet/docs#17297

We've updated our samples and snippets structure. This PR describes the new structure, and migration guidelines for existing snippets. In addition, rework the samples section to include the metadata necessary for the samples browser experience.

/cc @Thraka

### CODE REVIEW

File: Contribute/dotnet/dotnet-contribute.md [hunk 1]
```diff
@@ -77,6 +81,23 @@ Be sure to follow the proper Markdown syntax. For examples of common , see the [
...
+              porting-overview.fs
+              Program.fs
+            /vb
+              porting.vbproj
+              porting-overview.vb
+              Program.vb
+
+The structure shown above includes one image, *portability_report.png*, and three code projects that include **code snippets** that are included in the *porting-overview.md* article.
+
+For historical reasons, many of hte included snippets are stored under the */samples* folder in the *dotnet/docs* repository. If you are making major changes to an article, those snippets should be moved to the new structure. Do not move snippets for small changes.
```
Comment (tdykstra) [inline Contribute/dotnet/dotnet-contribute.md:100]:
```suggestion
For historical reasons, many of the included snippets are stored under the */samples* folder in the *dotnet/docs* repository. If you are making major changes to an article, those snippets should be moved to the new structure. Do not move snippets for small changes.
```

File: Contribute/dotnet/dotnet-contribute.md [hunk 2]
```diff
@@ -173,7 +192,23 @@ You build any .NET Core snippet or sample using the .NET Core CLI, which can be
 
 3. Add a readme.md to the root directory of your sample.
 
-   This should include a brief description of the code, and refer people to the article that references the sample.
+   This should include a brief description of the code, and refer people to the article that references the sample. The top of the readme.md must have the metadata required for the [samples browser](https://docs.microsoft.com/samples). The header block should contain the following fields:
```
Comment (tdykstra) [inline Contribute/dotnet/dotnet-contribute.md:195]:
```suggestion
   This should include a brief description of the code, and refer people to the article that references the sample. The top of the *readme.md* must have the metadata required for the [samples browser](https://docs.microsoft.com/samples). The header block should contain the following fields:
```

File: Contribute/dotnet/dotnet-contribute.md [hunk 3]
```diff
@@ -173,7 +192,23 @@ You build any .NET Core snippet or sample using the .NET Core CLI, which can be
...
+   ---
+   name: "really cool sample"
+   description: "Learn everything about this really cool sample."
+   page_type: sample
+   languages:
+     - csharp
+     - fsharp
+     - vbnet
+   products:
+     - dotnet-core
```
Comment (tdykstra) [inline Contribute/dotnet/dotnet-contribute.md:207]: Are there other products that samples might be for, that we should list identifiers for?

File: Contribute/dotnet/dotnet-contribute.md [hunk 4]
```diff
@@ -120,7 +139,7 @@ Each complete sample that you create should contain a *readme.md* file. This fil
 
 Your topic will also contain links to the sample. Link directly to the sample's folder on GitHub.
 
-### Writing a new snippet or sample
+### Writing a new sample
```
Comment (tdykstra) [inline Contribute/dotnet/dotnet-contribute.md:142]: Can we link to a representative sample that illustrates what we're saying in this section?

**Comment by BillWagner** [comment]
@tdykstra Do you know who needs to approve this before merging? Or was there a issue with recording your review?

File: Contribute/dotnet/dotnet-contribute.md [hunk 5]
```diff
@@ -77,6 +81,23 @@ Be sure to follow the proper Markdown syntax. For examples of common , see the [
...
+            /fsharp
+              porting.fsproj
+              porting-overview.fs
+              Program.fs
+            /vb
+              porting.vbproj
+              porting-overview.vb
+              Program.vb
+
+The structure shown above includes one image, *portability_report.png*, and three code projects that include **code snippets** that are included in the *porting-overview.md* article.
```
Comment (Thraka) [inline Contribute/dotnet/dotnet-contribute.md:98]: I think this should show more of the standard structure. This is assuming that this article is the only one in the folder.

Comment (tdykstra) [reply Contribute/dotnet/dotnet-contribute.md:98]: That doesn't look right.  Should be:

```
/snippets
  /porting-overview
    /csharp
       porting.csproj
        porting-overview.cs
        Program.cs
    /fsharp
      porting.fsproj
      porting-overview.fs
      Program.fs
    /vb
      porting.vbproj
      porting-overview.vb
      Program.vb
```

Comment (tdykstra) [reply Contribute/dotnet/dotnet-contribute.md:98]: >  Do you know who needs to approve this before merging? Or was there a issue with recording your review?

When it's ready to go, add the `#sign-off` comment and tag Jim Parker.

Comment (BillWagner) [reply Contribute/dotnet/dotnet-contribute.md:98]: I thought we'd agreed on the structure @tdykstra added in his comment. I'll make that update.

**Comment by BillWagner** [comment]
#sign-off

/cc @Jim-Parker

File: Contribute/dotnet/dotnet-contribute.md [hunk 6]
```diff
@@ -82,6 +82,20 @@ Be sure to follow the proper Markdown syntax. For examples of common , see the [
...
+                porting-overview.cs
+                Program.cs
+              /fsharp
+                porting.fsproj
+                porting-overview.fs
+                Program.fs
+               /vb
+                 porting.vbproj
+                 porting-overview.vb
+          Program.vb
```
Comment (Thraka) [inline Contribute/dotnet/dotnet-contribute.md:97]:
```suggestion
                 Program.vb
```

File: Contribute/dotnet/dotnet-contribute.md [hunk 7]
```diff
@@ -82,6 +82,20 @@ Be sure to follow the proper Markdown syntax. For examples of common , see the [
...
+              /snippets
```
Comment (Thraka) [inline Contribute/dotnet/dotnet-contribute.md:98]: I'm not sure what this?

Comment (BillWagner) [reply Contribute/dotnet/dotnet-contribute.md]: copy / paste error. Fixed.

File: Contribute/dotnet/dotnet-contribute.md [hunk 8]
```diff
@@ -63,7 +63,11 @@ For small changes, see the instructions on editing in GitHub on the [home page](
...
+
+For images and other static resources, create a subfolder called **media** inside the folder that contains your article, if it doesn't already exist. Inside the **media** folder, create a subfolder with the article name (except for the index file). 
+
+For **code snippets**, create a subfolder called **snippets** inside the folder that contains your article, if it doesn't already exist. In most cases, you'll have code snippets for all three of the main .NET languages, C#, F#, and Visual Basic. In that case, create sub-folders named **csharp**, **fsharp**, and **vb** for each of the three projects. For simplicity, use the **snippets** folder for your project in the C# Guide, F# Guide, and Visual Basic Guide. Those areas typically have snippets for one language. Code snippets are small, focused examples of code that demonstrate the concepts covered in an article. Larger programs, intended for download and exploration should be located in the [dotnet/samples](https://github.com/dotnet/samples) repository. Full samples are covered in the section on [Contributing to samples](#contributing-to-samples).
```
Comment (Thraka) [inline Contribute/dotnet/dotnet-contribute.md:70]: So it's easier to see, this suggestion changes the following:

> For code snippets, create a subfolder called snippets inside the folder that contains your article, if it doesn't already exist. **Inside the snippets folder, create a subfolder with the article name.** In most cases, you'll have code snippets for all three of the main 

```suggestion
For **code snippets**, create a subfolder called **snippets** inside the folder that contains your article, if it doesn't already exist. Inside the **snippets** folder, create a subfolder with the article name. In most cases, you'll have code snippets for all three of the main .NET languages, C#, F#, and Visual Basic. In that case, create sub-folders named **csharp**, **fsharp**, and **vb** for each of the three projects. For simplicity, use the **snippets** folder for your project in the C# Guide, F# Guide, and Visual Basic Guide. Those areas typically have snippets for one language. Code snippets are small, focused examples of code that demonstrate the concepts covered in an article. Larger programs, intended for download and exploration should be located in the [dotnet/samples](https://github.com/dotnet/samples) repository. Full samples are covered in the section on [Contributing to samples](#contributing-to-samples).
```

**Pull request merged**